src/rust/bin/wallet_tool.rs

```diff
@@ -0,0 +1,580 @@
+    }
+    let phrase = phrase_line[0]
+        .as_ref()
+        .unwrap()
+        .trim_start_matches("# - recovery_phrase=\"")
+        .trim_end_matches('"');
+
+    // This panic hook allows us to make a best effort to clear the screen (and then print
+    // another reminder about secrets in the export file) even if a panic occurs.
+
+    let old_hook = panic::take_hook();
+    {
+        let export_path = export_path.to_string();
+        panic::set_hook(Box::new(move |panic_info| {
+            clear_and_show_cautions(&export_path);
+
+            let s = panic_info.payload().downcast_ref::<&str>().unwrap_or(&"");
+            eprintln!("\nPanic: {}\n{:?}", s, Backtrace::new());
+        }));
+    }
+
+    let res = (|| -> anyhow::Result<()> {
+        println!("The recovery phrase is:\n");
+
+        const WORDS_PER_LINE: usize = 3;
+
+        let words: Vec<_> = phrase.split(' ').collect();
+        let max_len = words.iter().map(|w| w.len()).max().unwrap_or(0);
+
+        for (i, word) in words.iter().enumerate() {
+            print!("{0:2}: {1:2$}", i + 1, word, max_len + 2);
+            if (i + 1) % WORDS_PER_LINE == 0 {
+                println!();
+            }
+        }
+        if words.len() % WORDS_PER_LINE != 0 {
+            println!();
+        }
+
+        println!(concat!(
+            "\nPlease write down this phrase on something durable that you will keep\n",
+            "in a secure location.\n",
+            "Press Enter when finished; then the phrase will disappear and you'll be\n",
+            "asked to re-enter a selection of words from it."
+        ));
+
+        let mut stdin = io::stdin();
+        let mut buf = String::new();
+        prompt(&mut stdin, &mut buf)?;
+
+        // The only reliable and portable way to make sure the recovery phrase
+        // is no longer displayed is to clear the whole terminal (including
+        // scrollback, if possible). The text is only printed if clearing fails.
+        try_to_clear(concat!(
+            "\n\n\n\n\n\n\n\n\n\n\n\n",
+            "Please adjust the terminal window so that you can't see the\n",
+            "recovery phrase above. After finishing the backup, close the\n",
+            "terminal window or clear it"
+        ));
+
+        println!("\nNow we're going to confirm that you backed up the recovery phrase.");
+
+        let mut rng = thread_rng();
+        let mut unconfirmed: Vec<usize> = (0..words.len()).collect();
+        for _ in 0..min(3, words.len()) {
+            let index: usize = rng.gen_range(0..unconfirmed.len());
+            let n = unconfirmed[index];
+            unconfirmed[index] = unconfirmed[unconfirmed.len() - 1];
+            unconfirmed.pop().expect("should be nonempty");
+
+            loop {
+                let mut buf = String::new();
+                println!("\nPlease enter the {} word:", ordinal(n + 1));
+                let line = prompt(&mut stdin, &mut buf)?;
```

Comment:
This string is being used to store the entered words in memory, and is not being zeroized afterwards; this leaves the words on the heap. Use `secrecy::SecretString` instead. We also want to avoid re-allocations due to growing `String`; pre-allocate the maximum buffer size we expect.

The `prompt` API complicates this because it attempts to perform both line-reading and trimming, so it takes a mutable buffer which `SecretString` doesn't allow. Separate out the trimming.

Reply:
The secret words are necessarily printed to the terminal, so it isn't possible to ensure they are zeroized in all the places they occur in memory. That would depend on the implementation of at least the I/O libraries in Rust and C, and the implementation of the terminal and its clearing function. After careful consideration and trying to implement part of it, I think that zeroizing just this copy is a lot of work for little gain in security.

Reply:
We paired on this and it turned out to not be a lot of work.

Reply:
@str4d showed me this was easier than I thought; it's resolved partially now (the remainder being #5650).